src/Razor/src/Microsoft.VisualStudio.Editor.Razor/Snippets/XmlSnippetParser.ExpansionTemplate.cs

```diff
@@ -0,0 +1,233 @@
+                                    // LSP client currently only invokes on typing (tab) so there is no way to have a selection as part of a snippet request.
+                                    // Additionally, TM_SELECTED_TEXT is not supported in the VS LSP client, so we can't set the selection even if we wanted to.
+                                    // Since there's no way for the user to ask for a selection replacement, we can ignore it.
+                                }
+                                else if (string.Equals(fieldName, End, StringComparison.Ordinal))
+                                {
+                                    snippetParts.Add(new SnippetCursorPart());
+                                }
+                                else if (string.Equals(fieldName, Shortcut, StringComparison.Ordinal))
+                                {
+                                    snippetParts.Add(new SnippetShortcutPart());
+                                }
+                                else
+                                {
+                                    var field = FindField(fieldName);
+                                    if (field != null)
+                                    {
+                                        // If we have an editable field we need to know its order in the snippet so we can place the appropriate tab stop indices.
+                                        int? fieldIndex = field.IsEditable ? fieldNameToSnippetIndex.GetOrAdd(field.ID, (key) => currentTabStopIndex++) : null;
+                                        var fieldPart = string.IsNullOrEmpty(field.FunctionName)
+                                                    ? new SnippetFieldPart(field.ID, field.Default, fieldIndex)
+                                                    : new SnippetFunctionPart(field.ID, field.Default, fieldIndex, field.FunctionName!, field.FunctionParam);
+                                        snippetParts.Add(fieldPart);
+                                    }
+                                }
+                            }
+                            else
+                            {
+                                // simply append a '$'    
+                                snippetParts.Add(new SnippetStringPart(_delimiter));
+                            }
+
+                            // start the new token at the next character
+                            currentTokenCharIndex = currentCharIndex;
+                            currentTokenCharIndex++;
+                        }
+
+                        break;
+                }
+
+                currentCharIndex++;
+            }
+
+            // do we have any remaining text to be copied?
+            if (sps == SnippetParseState.Code && (currentCharIndex > currentTokenCharIndex))
+            {
+                var remaining = _code[currentTokenCharIndex..currentCharIndex];
+                snippetParts.Add(new SnippetStringPart(remaining));
```

Comment:
Consider holding onto a `ReadOnlyMemory<char>` that references the original _code rather that allocating a new string.